opm/material/fluidsystems/blackoilpvt/GasPvtThermal.hpp

```diff
@@ -222,16 +241,61 @@ class GasPvtThermal
                               const Evaluation& temperature,
-                              const Evaluation&,
-                              const Evaluation&) const
+                              const Evaluation& pressure,
+                              const Evaluation& Rv) const
     {
         if (!enableInternalEnergy_)
-            throw std::runtime_error("Requested the internal energy of oil but it is disabled");
+             throw std::runtime_error("Requested the internal energy of gas but it is disabled");
 
-        // compute the specific internal energy for the specified tempature. We use linear
-        // interpolation here despite the fact that the underlying heat capacities are
-        // piecewise linear (which leads to a quadratic function)
-        return internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true);
+        bool enableJouleThomson_  = gasJT_[regionIdx];
+
+        if (!enableJouleThomson_) {
+            // compute the specific internal energy for the specified tempature. We use linear
+            // interpolation here despite the fact that the underlying heat capacities are
+            // piecewise linear (which leads to a quadratic function)
+            return internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true);
+        }
+        else {
+            Evaluation Tref = gasdentRefTemp_[regionIdx];
+            Evaluation Pref = gasJTRefPres_[regionIdx]; 
+            Scalar JTC = gasJTC_[regionIdx]; // JTC = 0: JTC calculated
+
+            Evaluation invB = inverseFormationVolumeFactor(regionIdx, temperature, pressure, Rv);
+            const Scalar hVap = 480.6e3; // [J / kg]
+            Evaluation Cp = (internalEnergyCurves_[regionIdx].eval(temperature, /*extrapolate=*/true) - hVap)/temperature;
+            Evaluation density = invB * gasReferenceDensity(regionIdx);
+
+            Evaluation enthalpyPres;
+            if  (JTC != 0) {
+                enthalpyPres = -Cp * JTC * (pressure -Pref);
+            }
+            else if(enableThermalDensity_) {
+                Scalar c1T = gasdentCT1_[regionIdx];
+                Scalar c2T = gasdentCT2_[regionIdx];
+
+                Evaluation alpha = (c1T + 2 * c2T * (temperature - Tref)) /
+                    (1 + c1T  *(temperature - Tref) + c2T * (temperature - Tref) * (temperature - Tref));
+
+                const int N = 100; // value is experimental
+                Evaluation deltaP = (pressure - Pref)/N;
+                Evaluation enthalpyPresPrev = 0;
+                for (size_t i = 0; i < N; ++i) {
+                    Evaluation Pnew = Pref + i * deltaP;
+                    Evaluation rho = inverseFormationVolumeFactor(regionIdx, temperature, Pnew, Rv) * gasReferenceDensity(regionIdx);
+                    Evaluation jouleThomsonCoefficient = -(1.0/Cp) * (1.0 - alpha * temperature)/rho;  
+                    Evaluation deltaEnthalpyPres = -Cp * jouleThomsonCoefficient * deltaP;
```

Comment:
Is the division by rho here to get the specific heat-capacity ? cp vs Cp.

Doesn't rho here need to take into account the Rv?
https://github.com/OPM/opm-material/blob/9eb5c71a2f534e80d2ef30f58708f7a49464058e/opm/material/fluidsystems/BlackOilFluidSystem.hpp#L811

I guess the same comment apply for the other density calculations.